dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java

```diff
@@ -368,6 +374,10 @@ public void testRefreshToken() throws Exception {
 
         String newToken = getClient(token).perform(post("/api/authn/login"))
                                           .andExpect(status().isOk())
+                                          // An auth token refresh should also refresh the CSRF token
+                                          // (which should be returned in both a cookie & header).
+                                          .andExpect(cookie().exists("DSPACE-XSRF-COOKIE"))
+                                          .andExpect(header().exists("DSPACE-XSRF-TOKEN"))
```

Comment:
it would be nice if we can check that the token is the same in both place

Reply:
Good point. I can add a line to check that the cookie() matches the header().  It always should, but it's worth checking.

Reply:
@abollini : On second thought, this is much harder than it seems :(  There's no easy way in Spring to check that that value returned in a Cookie is the same as the one in a Header.  

Also, we've modeled the behavior of Spring Security's `CsrfAuthenticationStrategy` code, which technically sets the Cookie *twice* in one response...first to an empty string ("") to clear it out, then to the new value (see https://github.com/spring-projects/spring-security/blob/master/web/src/main/java/org/springframework/security/web/csrf/CsrfAuthenticationStrategy.java#L58-L60).  So, the response technically looks like:
```
Set-Cookie: DSPACE-XSRF-COOKIE=; [other settings]
Set-Cookie: DSPACE-XSRF-COOKIE=[new-token]; [other settings]
DSPACE-XSRF-TOKEN: [new-token]
```

Spring's `cookie()` matcher seems to assume you only have one Cookie of a name, and there's no way to tell `cookie().value()` to match the *second* value and not the *first* one (which will always be an empty string when a token change occurs).

In other words, I think this test might be as good as we can get without enhancing the test framework overall (which would take more time than we currently have).  We already have proven (in other tests) that this Cookie & Header are only returned when we expect a CSRF token change to occur.

Reply:
The best I've come up with in terms of *verifying* the CSRF Token, is something like this:
```
.andExpect(cookie().exists("DSPACE-XSRF-COOKIE"))
.andExpect(header().exists("DSPACE-XSRF-TOKEN"))
// First cookie will always be empty. Second has the token & there's no way to match it easily
.andExpect(cookie().value("DSPACE-XSRF-COOKIE", ""))
// CSRF Tokens generated by Spring Security are UUIDs
.andExpect(header().string("DSPACE-XSRF-TOKEN", matchesPattern(REGEX_UUID)));
```

This isn't what you've asked for, but it's slightly better than just checking for existence.  If you feel this is better, I can add it to the tests here.

Reply:
thanks @tdonohue this is absolutely ok as it tests exactly what you (we) expect to happen. I'm not sure that it is a good idea to add an empty cookie before as the new one should already update any previous information on the browser side but as I was not able to find any specs that prohibits to do that and there are no evidence of side effect I'm ok with that

Reply:
Just in case we are wondering why the empty cookie approach is being used.  Again, I chose that approach as it's the same approach used by Spring Security in their `CsrfAuthenticationStrategy`:
1. First, they [set the CSRF token to `null`](https://github.com/DSpace/dspace-angular/issues/1069), which creates an empty cookie first
2. Then, they [set the new Token](https://github.com/spring-projects/spring-security/blob/5.4.x/web/src/main/java/org/springframework/security/web/csrf/CsrfAuthenticationStrategy.java#L60), which creates a second cookie with the token in it.

I'll make sure to document this as well in the comments of the test itself (which I'll create shortly)